src/main/java/com/google/devtools/build/lib/worker/WorkerMultiplexer.java

```diff
@@ -0,0 +1,211 @@
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.worker;
+
+import com.google.devtools.build.lib.shell.Subprocess;
+import com.google.devtools.build.lib.shell.SubprocessBuilder;
+import com.google.devtools.build.lib.worker.WorkerProtocol.WorkResponse;
+import com.google.devtools.build.lib.vfs.Path;
+import java.io.File;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Semaphore;
+
+/** An intermediate worker that sends request and receives response from the worker processes. */
+public class WorkerMultiplexer extends Thread {
+  /**
+   * WorkerMultiplexer is running as a thread on its own. When worker process
+   * returns the WorkResponse, it is stored in this map and wait for
+   * WorkerProxy to retrieve the response.
+   */
+  private Map<Integer, InputStream> workerProcessResponse;
+  /**
+   * A semaphore to protect workerProcessResponse object.
+   */
+  private Semaphore semWorkerProcessResponse;
+  /**
+   * After sending the WorkRequest, WorkerProxy will wait on a semaphore to be
+   * released. WorkerMultiplexer is responsible to release the corresponding
+   * semaphore in order to signal WorkerProxy that the WorkerResponse has been
+   * received.
+   */
+  private Map<Integer, Semaphore> responseChecker;
+  /**
+   * A semaphore to protect responseChecker object.
+   */
+  private Semaphore semResponseChecker;
+  /**
+   * The worker process that this WorkerMultiplexer should be talking to.
+   */
+  private Subprocess process;
+
+  private Thread shutdownHook;
+  private Integer workerHash;
+
+  WorkerMultiplexer(Integer workerHash) {
+    semWorkerProcessResponse = new Semaphore(1);
+    semResponseChecker = new Semaphore(1);
+    responseChecker = new HashMap<>();
+    workerProcessResponse = new HashMap<>();
+    this.workerHash = workerHash;
+
+    final WorkerMultiplexer self = this;
+    this.shutdownHook =
+      new Thread(
+        () -> {
+          try {
+            self.shutdownHook = null;
+            self.destroyMultiplexer();
+          } finally {
+            // We can't do anything here.
+          }
+        });
+    Runtime.getRuntime().addShutdownHook(shutdownHook);
+  }
+
+  /**
+   * Only start one worker process for each WorkerMultiplexer, if it hasn't.
+   */
+  public synchronized void createProcess(WorkerKey workerKey, Path workDir, Path logFile) throws IOException {
```

Comment:
I might be missing something, but I think you can just make this class extend `Worker` and eliminate all of the duplicate logic by passing these arguments (which `Worker` expects in the constructor) to `WorkerMultiplexerManager.getInstance` rather than thread them through here.

Reply:
It is indeed duplicate code from `Worker`. I am also happy to find a way to avoid duplicate code, but I don't think making this class extend `Worker` is a good way.

`Worker` and `WorkerMultiplexer` serve different purpose. Process creation is the only part they share, but they still have a lot of distinct functions. `WorkerProxy` should be the one that extends `Worker` in my opinion. In addition, we have to prevent process creation if previous `WorkerProxy` has already done, so the `createProcess`es are not exactly the same.

Let me know if you come up a better way to avoid duplicates.